### PR TITLE
[WIP] Port forwarding using client-go

### LINKED ISSUE #1536: Port forward in monitoring should log errors
/area test-and-release

## Expected Behavior
The code that performs the port forwarding doesn't log any info. We need it in order to diagnose https://github.com/knative/eventing/issues/3570

## Actual Behavior
Nothing is logged from `kubectl port-forward` process started in background

## Additional Info
We might think to use directly the client-go to perform port forwarding https://godoc.org/k8s.io/client-go/tools/portforward#New

### PR DESCRIPTION
**Pull request by slinkydeveloper**
Removed code to perform port forwarding using an external process and replaced with code to perform port forwarding directly with client-go, included with proper shutdown and logging

Fixes #1536

### CODE REVIEW

File: test/monitoring/monitoring.go [hunk 1]
```diff
@@ -52,34 +57,39 @@ func GetPods(kubeClientset *kubernetes.Clientset, app, namespace string) (*v1.Po
...
-	portFwdCmd := fmt.Sprintf("kubectl port-forward %s %d:%d -n %s", podName, localPort, remotePort, namespace)
-	portFwdProcess, err := executeCmdBackground(logf, portFwdCmd)
-
-	if err != nil {
-		return 0, fmt.Errorf("failed to port forward: %w", err)
+// To close the port forwarding, just close the channel
+func PortForward(logf logging.FormatLogger, config *rest.Config, clientSet *kubernetes.Clientset, pod *v1.Pod, localPort, remotePort int) (chan struct{}, error) {
+	req := clientSet.RESTClient().Post().Resource("pods").Namespace(pod.Namespace).Name(pod.Name).SubResource("portforward")
+	portForwardUrl := req.URL()
+	if !strings.HasPrefix(portForwardUrl.Path, "/api/v1") {
```
Comment (evankanderson) [inline test/monitoring/monitoring.go:65]: ??? This seems like it deserves a comment at least.

File: test/monitoring/monitoring.go [hunk 2]
```diff
@@ -52,34 +57,39 @@ func GetPods(kubeClientset *kubernetes.Clientset, app, namespace string) (*v1.Po
...
 // PortForward sets up local port forward to the pod specified by the "app" label in the given namespace
-func PortForward(logf logging.FormatLogger, podList *v1.PodList, localPort, remotePort int, namespace string) (int, error) {
-	podName := podList.Items[0].Name
-	portFwdCmd := fmt.Sprintf("kubectl port-forward %s %d:%d -n %s", podName, localPort, remotePort, namespace)
-	portFwdProcess, err := executeCmdBackground(logf, portFwdCmd)
-
-	if err != nil {
-		return 0, fmt.Errorf("failed to port forward: %w", err)
+// To close the port forwarding, just close the channel
+func PortForward(logf logging.FormatLogger, config *rest.Config, clientSet *kubernetes.Clientset, pod *v1.Pod, localPort, remotePort int) (chan struct{}, error) {
```
Comment (evankanderson) [inline test/monitoring/monitoring.go:62]: The previous version took a *v1.PodList". (I also wonder if passing the logger, clientSet, etc would be more cleanly done via injection/`context.Context`.)

File: test/logging/logger_writer.go [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+package logging
+
+import (
+	"bytes"
+	"io"
+	"strings"
+	"sync"
+)
+
+type loggerWriter struct {
```
Comment (evankanderson) [inline test/logging/logger_writer.go:25]: Should these be "prefixedWriter" or "prefixWriter"?

File: test/monitoring/monitoring.go [hunk 4]
```diff
@@ -52,34 +57,39 @@ func GetPods(kubeClientset *kubernetes.Clientset, app, namespace string) (*v1.Po
...
 
-// RunBackground starts a background process and returns the Process if succeed
-func executeCmdBackground(logf logging.FormatLogger, format string, args ...interface{}) (*os.Process, error) {
-	cmd := fmt.Sprintf(format, args...)
-	logf("Executing command: %s", cmd)
-	parts := strings.Split(cmd, " ")
-	c := exec.Command(parts[0], parts[1:]...) // #nosec
-	if err := c.Start(); err != nil {
-		return nil, fmt.Errorf("%s command failed: %w", cmd, err)
+	transport, upgrader, err := spdy.RoundTripperFor(config)
```
Comment (evankanderson) [inline test/monitoring/monitoring.go:72]: Wow, this really is SPDY: https://github.com/kubernetes/kubernetes/issues/89163

File: test/logging/logger_writer.go (hunk 3 above)
Comment (slinkydeveloper) [reply test/logging/logger_writer.go:25]: My idea to name it `loggerWriter` is to clarify that you build this writer from a logger function

**Comment by slinkydeveloper** [comment]
> I'm assuming that overall we expect this to be somewhat faster with better cleanup because we aren't managing processes?

That's the point plus we have the logging

File: test/monitoring/monitoring.go (hunk 4 above)
Comment (slinkydeveloper) [reply test/monitoring/monitoring.go:72]: yeah :) i took that code straight from kubectl

File: test/monitoring/monitoring.go (hunk 2 above)
Comment (slinkydeveloper) [reply test/monitoring/monitoring.go:62]: That really didn't made sense, because you port forward one pod at the time

**Pull request closed without merging**